src/openms/source/ANALYSIS/OPENSWATH/OpenSwathScoring.cpp

```diff
@@ -532,92 +560,106 @@ namespace OpenMS
+  std::vector<OpenSwath::SpectrumPtr> OpenSwathScoring::fetchSpectrumSwath(OpenSwath::SpectrumAccessPtr swathmap, double RT, int nr_spectra_to_add, double drift_lower, double drift_upper)
   {
-    return getAddedSpectra_(std::move(swath_map), RT, nr_spectra_to_add, drift_lower, drift_upper);
+
+    std::vector<OpenSwath::SpectrumPtr> all_spectra = fetchMultipleSpectra_(swathmap, RT, nr_spectra_to_add);
+    if (spectra_addition_method_ == "simple")
+    {
+      return all_spectra;
+    }
+    else
+    {
+      std::vector<OpenSwath::SpectrumPtr> spectrum_out;
+      spectrum_out.push_back(getAddedSpectra_(all_spectra, drift_lower, drift_upper));
+      return spectrum_out;
+    }
   }
 
-  OpenSwath::SpectrumPtr OpenSwathScoring::fetchSpectrumSwath(std::vector<OpenSwath::SwathMap> swath_maps,
-                                                              double RT, int nr_spectra_to_add, const double drift_lower, const double drift_upper)
+  std::vector<OpenSwath::SpectrumPtr> OpenSwathScoring::fetchSpectrumSwath(std::vector<OpenSwath::SwathMap> swath_maps, double RT, int nr_spectra_to_add, double drift_lower, double drift_upper)
   {
+    OPENMS_PRECONDITION(nr_spectra_to_add >= 1, "nr_spectra_to_add must be at least 1.")
+    OPENMS_PRECONDITION(!swath_maps.empty(), "swath_maps vector cannot be empty")
+
+    // This is not SONAR data
     if (swath_maps.size() == 1)
     {
-      return getAddedSpectra_(swath_maps[0].sptr, RT, nr_spectra_to_add, drift_lower, drift_upper);
+      return fetchSpectrumSwath(swath_maps[0].sptr, RT, nr_spectra_to_add, drift_lower, drift_upper);
     }
     else
     {
-      // multiple SWATH maps for a single precursor -> this is SONAR data
+      // multiple SWATH maps for a single precursor -> this is SONAR data, in all cases only return a single spectrum
       std::vector<OpenSwath::SpectrumPtr> all_spectra;
       for (size_t i = 0; i < swath_maps.size(); ++i)
       {
-        OpenSwath::SpectrumPtr spec = getAddedSpectra_(swath_maps[i].sptr, RT, nr_spectra_to_add, drift_lower, drift_upper);
+        std::vector<OpenSwath::SpectrumPtr> spectrum_vector = fetchMultipleSpectra_(swath_maps[i].sptr, RT, nr_spectra_to_add);
+        OpenSwath::SpectrumPtr spec = getAddedSpectra_(spectrum_vector, drift_lower, drift_upper);
         all_spectra.push_back(spec);
       }
       OpenSwath::SpectrumPtr spectrum_ = SpectrumAddition::addUpSpectra(all_spectra, spacing_for_spectra_resampling_, true);
-      return spectrum_;
+      std::vector<OpenSwath::SpectrumPtr> spectrum_out;
+      spectrum_out.push_back(spectrum_);
+      return spectrum_out;
     }
   }
 
-  OpenSwath::SpectrumPtr filterByDrift(const OpenSwath::SpectrumPtr& input, const double drift_lower, const double drift_upper)
+  OpenSwath::SpectrumPtr OpenSwathScoring::filterByDrift_(const OpenSwath::SpectrumPtr& input, const double drift_lower, const double drift_upper)
```

Comment:
IMHO this function does not belong here.
Its a data manipulation function, and has little to do with scoring.
Hence, its hard to find and reuse.
we already have `IONMOBILITY/IMDataConverter.cpp`. Maybe this folder is a better place for this kind if function, so its more accessible to everyone.

Also, this function can be `static` (or a free function)

Reply:
Good point. Thought about this however since this uses OpenSwath specific data types I don't think it is appropriate to place in IONMOBILITY/IMDATACONVERTER. Perhaps this can be moved to DIAHelpers though since it is more of a helpter function than a scoring function

Reply:
DIAHelpers only improves the visibility by a small margin.
If you feel it does not belong in IONMOBILITY then create a new IONMOBILITY subfolder in openswath and put there.

Reply:
I'm not a huge fan of creating a new IONMOBILITY subfolder yet, because I think in the long run it is better to make the downstream functions "ion mobility aware" Currently I'm thinking of placing it in `openswathalgo/dataaccess/iSpectrumAccess` with the idea that when the spectrum is fetched it can be filtered right away.

Reply:
However this is in the openswathalgo folder which leads to some limitations